### PR TITLE
feat: add REST API handlers and client methods for user secrets

### PR DESCRIPTION
**Pull request by zedkipp**
Adds the five REST endpoints for managing user secrets, SDK client methods, and handler tests.

Endpoints:
- `POST /api/v2/users/{user}/secrets`
- `GET /api/v2/users/{user}/secrets`
- `GET /api/v2/users/{user}/secrets/{name}`
- `PATCH /api/v2/users/{user}/secrets/{name}`
- `DELETE /api/v2/users/{user}/secrets/{name}`

Routes are registered under the existing `/{user}` group with `ExtractUserParam`. The delete query was changed from `:exec` to `:execrows` so the handler can distinguish "not found" from success (DELETE with `:exec` silently returns nil for zero affected rows).

Depends on #24102 and #24103.

### CODE REVIEW

**Comment by zedkipp** [comment]
* **#24107** <a href="https://app.graphite.com/github/pr/coder/coder/24107?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/coder/coder/24107?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#24103** <a href="https://app.graphite.com/github/pr/coder/coder/24103?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24102** <a href="https://app.graphite.com/github/pr/coder/coder/24102?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: coderd/usersecrets_test.go [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+		})
+		require.Error(t, err)
+		var sdkErr *codersdk.Error
+		require.ErrorAs(t, err, &sdkErr)
+		assert.Equal(t, http.StatusBadRequest, sdkErr.StatusCode())
+	})
+
+	t.Run("MissingValue", func(t *testing.T) {
+		t.Parallel()
+		client := coderdtest.New(t, nil)
```
Comment (zedkipp) [inline coderd/usersecrets_test.go]: Should avoid creating a whole new server and ancillaries for every sub-test.

**Comment by zedkipp** [comment]
I'm ignoring the docs check bot comment because the feature doesn't exist as a whole yet. We're laying groundwork here. We'll do feature docs updates when this gets closer to shipping.

File: coderd/usersecrets.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+// @Param name path string true "Secret name"
+// @Success 200 {object} codersdk.UserSecret
+// @Router /users/{user}/secrets/{name} [get]
+func (api *API) getUserSecret(rw http.ResponseWriter, r *http.Request) { //nolint:revive // Method name matches route.
+	ctx := r.Context()
+	user := httpmw.UserParam(r)
+	name := chi.URLParam(r, "name")
+
+	secret, err := api.Database.GetUserSecretByUserIDAndName(ctx, database.GetUserSecretByUserIDAndNameParams{
+		UserID: user.ID,
```
Comment (dylanhuff-at-coder) [inline coderd/usersecrets.go:134]: Thinking out loud/question for my own understanding: 
An auth check would have already happened before we get into this method, or the others listed here? If an auth check was already done, then do we need to corroborate that the credentials used to auth correspond to the UserID passed in?

Comment (dylanhuff-at-coder) [reply coderd/usersecrets.go:134]: I figure there is an interceptor that already handles both of those concerns but want to double check :)

File: coderd/usersecrets.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+	secret, err := api.Database.UpdateUserSecretByUserIDAndName(ctx, params)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			httpapi.ResourceNotFound(rw)
+			return
+		}
+		if database.IsUniqueViolation(err) {
+			httpapi.Write(ctx, rw, http.StatusConflict, codersdk.Response{
+				Message: "Update would conflict with an existing secret.",
```
Comment (dylanhuff-at-coder) [inline coderd/usersecrets.go:235]: This is confusing to me, shouldn't patch require an existing secret?

File: coderd/usersecrets_test.go [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+		t.Parallel()
+		ctx := testutil.Context(t, testutil.WaitMedium)
+
+		_, err := client.CreateUserSecret(ctx, codersdk.Me, codersdk.CreateUserSecretRequest{
+			Value: "some-value",
+		})
+		require.Error(t, err)
+		var sdkErr *codersdk.Error
+		require.ErrorAs(t, err, &sdkErr)
+		assert.Equal(t, http.StatusBadRequest, sdkErr.StatusCode())
```
Comment (dylanhuff-at-coder) [inline coderd/usersecrets_test.go:50]: For this case and the missing value one, it maybe worth it to add an assertion that the error message specifies missing name or value

File: coderd/usersecrets.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+		return
+	}
+	if err := codersdk.UserSecretFilePathValid(req.FilePath); err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid file path.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
```
Comment (dylanhuff-at-coder) [inline coderd/usersecrets.go:66]: Should we check that either env var name or file path is set? I suppose someone could make a secret then later update the path or var name with a patch, but I lean towards requiring either of them at create

File: coderd/usersecrets_test.go [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+		_, err := client.CreateUserSecret(ctx, codersdk.Me, codersdk.CreateUserSecretRequest{
+			Name:     "bad-path-secret",
+			Value:    "value",
+			FilePath: "relative/path",
+		})
+		require.Error(t, err)
+		var sdkErr *codersdk.Error
+		require.ErrorAs(t, err, &sdkErr)
+		assert.Equal(t, http.StatusBadRequest, sdkErr.StatusCode())
+	})
```
Comment (dylanhuff-at-coder) [inline coderd/usersecrets_test.go:190]: Could add another test for checking duplicate file path

File: coderd/usersecrets.go (hunk 2 above)
Comment (zedkipp) [reply coderd/usersecrets.go:134]: Good questions, the authentication is handled in middleware that's registered where the new routes were added in coderd/coderd.go. The group that the `/{user}` endpoints are under have the middleware registered with `r.Use(apiKeyMiddleware)` here https://github.com/coder/coder/blob/1e40cea1991551777e0f28118af4842e53f68755/coderd/coderd.go#L1538

Then the authorization check happens in `GetUserSecretByUserIDAndName()`, which you can see in my previous PR here: https://github.com/coder/coder/pull/23998. You'll see we check if the user making the request has permissions for `rbac.ResourceUserSecret`.

Comment (dylanhuff-at-coder) [reply coderd/usersecrets.go:134]: Cool thank you for the explanation

File: coderd/usersecrets.go (hunk 3 above)
Comment (zedkipp) [reply coderd/usersecrets.go:235]: This will happen if a user updates the environment variable name or file path to conflict with an existing secret. This uniqueness is enforced in the database right now for our current scope.

Comment (dylanhuff-at-coder) [reply coderd/usersecrets.go:235]: Okay right, I think adding that the conflict is because of an an existing secret that has the same env var name or file path would help troubleshoot when this error message is received

File: coderd/usersecrets.go (hunk 5 above)
Comment (zedkipp) [reply coderd/usersecrets.go:66]: It seems very possible a user wants to just add their secret name/value and set the file path/env var later. Similarly, they can clear the env name/file path to easily “turn off” a secret without deleting it.

**Review by dylanhuff-at-coder — APPROVED** [review]
A couple of small comments but otherwise looks good

File: coderd/usersecrets_test.go (hunk 4 above)
Comment (zedkipp) [reply coderd/usersecrets_test.go:50]: Good call, done!

File: coderd/usersecrets_test.go (hunk 6 above)
Comment (zedkipp) [reply coderd/usersecrets_test.go:190]: Agreed, I also added a test for a file path conflict on update
